src/box/txn_limbo.c

```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
+	 */
+	if (req->replica_id == REPLICA_ID_NIL) {
+		if (req->type != IPROTO_RAFT_PROMOTE) {
+			say_error("%s. Zero replica_id detected",
+				  reject_str(req));
+			diag_set(ClientError, ER_UNSUPPORTED, "Replication",
+				 "synchronous requests with zero replica_id");
+			return -1;
+		}
+	}
+	if (req->replica_id != limbo->owner_id) {
+		/*
+		 * Incoming packets should esteem limbo owner,
+		 * if it doesn't match it means the sender
+		 * missed limbo owner migrations and is out of date.
+		 */
+		say_error("%s. Limbo owner mismatch, owner_id %u",
+			  reject_str(req), limbo->owner_id);
+		diag_set(ClientError, ER_SPLIT_BRAIN,
+			 "got a request from a foreign synchro queue owner");
+		return -1;
+	}
+
+	return 0;
+}
+
+/**
+ * A common filter for all synchro requests, checking that request operates
+ * over a valid lsn range.
+ */
+static int
+filter_queue_boundaries(struct txn_limbo *limbo,
+			const struct synchro_request *req)
+{
+	int64_t lsn = req->lsn;
+	/*
+	 * Easy case - processed LSN matches the new one which comes inside
+	 * request, everything is consistent.
+	 */
+	if (limbo->confirmed_lsn == lsn)
+		return 0;
+
+	/*
+	 * Explicit split brain situation. Request comes in with an old LSN
+	 * which we've already processed.
+	 */
+	if (limbo->confirmed_lsn > lsn) {
+		say_error("%s. confirmed lsn %lld > request lsn %lld",
+			  reject_str(req), (long long)limbo->confirmed_lsn,
+			  (long long)lsn);
+		diag_set(ClientError, ER_SPLIT_BRAIN,
+			 "got a request with lsn from an already "
+			 "processed range");
+		return -1;
+	}
+
+	/*
+	 * The last case requires a few subcases.
+	 */
+	assert(limbo->confirmed_lsn < lsn);
+
+	if (txn_limbo_is_empty(limbo)) {
+		/*
+		 * Transactions are rolled back already,
+		 * since the limbo is empty.
+		 */
+		say_error("%s. confirmed lsn %lld < request lsn %lld "
+			  "and empty limbo", reject_str(req),
+			  (long long)limbo->confirmed_lsn,
+			  (long long)lsn);
+		diag_set(ClientError, ER_SPLIT_BRAIN,
+			 "got a request mentioning future lsn");
+		return -1;
+	} else {
+		/*
+		 * Some entries are present in the limbo, we need to make sure
+		 * that request lsn lays inside limbo [first; last] range.
+		 * So that the request has some queued data to process,
+		 * otherwise it means the request comes from split brained node.
+		 */
+		int64_t first_lsn = txn_limbo_first_entry(limbo)->lsn;
+		int64_t last_lsn = txn_limbo_last_synchro_entry(limbo)->lsn;
+
+		if (lsn < first_lsn || last_lsn < lsn) {
+			say_error("%s. request lsn %lld out of range "
+				  "[%lld; %lld]", reject_str(req),
+				  (long long)lsn,
+				  (long long)first_lsn,
+				  (long long)last_lsn);
+			diag_set(ClientError, ER_SPLIT_BRAIN,
+				 "got a request lsn out of queue range");
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * Filter CONFIRM and ROLLBACK packets.
+ */
+static int
+filter_confirm_rollback(struct txn_limbo *limbo,
+			const struct synchro_request *req)
```

Comment:
No problem:
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index 73cddd596..bbdf1a983 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -947,7 +947,8 @@ txn_limbo_filter_confirm_rollback(struct txn_limbo *limbo,
 {
        assert(latch_is_locked(&limbo->promote_latch));
        assert(limbo->do_validate);
-
+       assert(req->type == IPROTO_RAFT_CONFIRM ||
+              req->type == IPROTO_RAFT_ROLLBACK);
        /*
         * Zero LSN is allowed for PROMOTE and DEMOTE requests only.
         */
@@ -968,7 +969,7 @@ txn_limbo_filter_promote_demote(struct txn_limbo *limbo,
 {
        assert(latch_is_locked(&limbo->promote_latch));
        assert(limbo->do_validate);
-
+       assert(iproto_type_is_promote_request(req->type));
        /*
         * PROMOTE and DEMOTE packets must not have zero
         * term supplied, otherwise it is a broken packet.
```